### PR TITLE
[json-schema] Added type definition for JSON object

### PR DESCRIPTION
**Pull request by JasonHK**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Review by SimonSchick — CHANGES_REQUESTED** [review]
You made a lot of unrelated style and formatting changes, please put undo those, put them in a separate commit or ideally a separate PR so us reviewers can focus on what has actually changed.

**Comment by JasonHK** [comment]
> You made a lot of unrelated style and formatting changes, please put undo those, put them in a separate commit or ideally a separate PR so us reviewers can focus on what has actually changed.

@SimonSchick Sorry about that, I've reverted the style changes.

**Comment by JasonHK** [comment]
> @SimonSchick Thank you for reviewing this PR! The author has pushed new commits since your last review. Could you take another look and submit a fresh review?
> 

@SimonSchick Could you please take a look at the updated PR?

**Comment by JasonHK** [comment]
@bcherny @cyrilletuzi @LucianBuzzo @rolandjitsu @SimonSchick Just a friendly reminder for you guys to review the PR.

File: types/json-schema/index.d.ts [hunk 1]
```diff
@@ -4,21 +4,46 @@
...
+    | string
+    | number
+    | boolean
+    | JSONSchema4Object
+    | JSONSchema4Array
+    | null;
+
+// Workaround for infinite type recursion
+export interface JSONSchema4Object {
+[key: string]: JSONSchema4Type;
```
Comment (rbuckton) [inline types/json-schema/index.d.ts:41]: nit: indentation

**Review by rbuckton — CHANGES_REQUESTED** [review]
Can you fix the indentation here? Then I can merge.

File: types/json-schema/index.d.ts (hunk 1 above)
Comment (JasonHK) [reply types/json-schema/index.d.ts]: Fixed

**Comment by orta** [comment]
Alright, this all looks good - let's get this in. Thanks @jasonHK for sticking with it!

**Pull request merged**